pkg/operator/certrotation/signer_test.go

```diff
@@ -19,6 +26,170 @@ import (
+		t.Fatalf("missing data in 'tls.crt' key of Data: %#v", existing.Data)
+	}
+	tlsKeyWant, ok := existing.Data["tls.key"]
+	if !ok || len(tlsKeyWant) == 0 {
+		t.Fatalf("missing data in 'tls.key' key of Data: %#v", existing.Data)
+	}
+
+	// copy the existing object before test begins, so we can diff it against
+	// the final object on the cluster after the controllers finish
+	secretWant := existing.DeepCopy()
+
+	indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	if err := indexer.Add(existing); err != nil {
+		t.Fatal(err)
+	}
+	clientset := kubefake.NewSimpleClientset(existing)
+
+	// the list cache is synced as soon as we have a delete, create, or update
+	clientset.PrependReactor("delete", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		if err = indexer.Delete(existing); err != nil {
+			t.Errorf("unexpected error while syncing the cache, op=delete: %v", err)
+		}
+		return false, nil, nil
+	})
+	clientset.PrependReactor("create", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		switch action := action.(type) {
+		case clienttesting.CreateActionImpl:
+			indexer.Delete(existing)
+			if err := indexer.Add(action.GetObject()); err != nil {
+				t.Errorf("unexpected error while syncing the cache, op=create: %v", err)
+			}
+			return false, action.GetObject(), nil
+		}
+		t.Errorf("wrong test setup")
+		return false, nil, nil
+	})
+	clientset.PrependReactor("update", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		switch action := action.(type) {
+		case clienttesting.UpdateActionImpl:
+			indexer.Delete(existing)
+			if err := indexer.Add(action.GetObject()); err != nil {
+				t.Errorf("unexpected error while syncing the cache, op=update: %v", err)
+			}
+			return false, action.GetObject(), nil
+		}
+		t.Errorf("wrong test setup")
+		return false, nil, nil
+	})
+
+	// r := rand.New(rand.NewSource(time.Now().Unix()))
+	options := events.RecommendedClusterSingletonCorrelatorOptions()
+	client := clientset.CoreV1().Secrets(ns)
+	var wg sync.WaitGroup
+	var id int32
+
+	controllers := 20
+	wg.Add(controllers)
+	for i := 1; i <= controllers; i++ {
+		go func() {
+			defer wg.Done()
+
+			controllerName := fmt.Sprintf("controller-%d", atomic.AddInt32(&id, 1))
+			recorder := events.NewKubeRecorderWithOptions(clientset.CoreV1().Events(ns), options, "operator", &corev1.ObjectReference{Name: controllerName, Namespace: ns})
+			wrapped := &wrapped{SecretInterface: client, name: controllerName, t: t}
+			getter := &getter{w: wrapped}
+			ctrl := &RotatedSigningCASecret{
+				Namespace:             ns,
+				Name:                  name,
+				Validity:              24 * time.Hour,
+				Refresh:               12 * time.Hour,
+				Client:                getter,
+				Lister:                corev1listers.NewSecretLister(indexer),
```

Comment:
Would it be simpler to provide a test implementation of SecretLister using the fake client's object tracker instead of using client reactor functions to simulate a reflector?

Reply:
FWIW the client reactor lets us have more control between a) object changes and b) the cache is updated, not that  we are exercising it in this test, but yeah i think we could use a fake `SecretLister` implementation.